src/main/java/backend/control/RepoOpControl.java

```diff
@@ -42,6 +64,26 @@ public RepoOpControl(RepoIO repoIO) {
         return result;
     }
 
+    /**
+     * Updates repository stored locally with data from a GitHubModelUpdatesData object.
+     * Set syncOperation to queue this operation in the blocking queue for the updating repository
+     * @param updates
+     * @param syncOperation
+     * @return
+     */
+    public CompletableFuture<Model> updateLocalModel(GitHubModelUpdatesData updates,
+                                                     boolean syncOperation) {
+        init(updates.getRepoId());
+        CompletableFuture<Model> result = new CompletableFuture<>();
+        UpdateLocalModelOp op = new UpdateLocalModelOp(models, updates, result);
+        if (syncOperation) {
```

Comment:
This is due to how several methods in `RepoIO` works. When opening the model, which is an operation placed on the queue, `RepoIO` will call `updateModel`, which is then executed on the thread for that queue. Then `updateModel` will place this `updateLocalModel` operation on the queue and wait for it to be executed. Consequently there will be a deadlock. So this operation allows to be in two modes, one of which is to bypass putting the operation on the queue if the executing thread is already the thread for that queue.

Now, we can divided this `updateModel` method in `RepoIO` into smaller methods that does each step such as `download` -> `updateLocalModel` -> `writeToJSON`. However, the `writeToJSON` part of the `updateModel` method includes a corrupted JSON explosion detection and recovery that needs to recursively call the whole chain of things again, making it quite hard to separate these steps. So that's the dilemma at the moment.